pkg/controller/endpointslice/metrics/metrics.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"sync"
+
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+// EndpointSliceSubsystem - subsystem name used for Endpoint Slices.
+const EndpointSliceSubsystem = "endpoint_slice_controller"
+
+var (
+	// EndpointSliceEndpointsAddedPerSync tracks the number of endpoints
+	// added on each Service sync.
+	EndpointSliceEndpointsAddedPerSync = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Subsystem:      EndpointSliceSubsystem,
+			Name:           "endpoints_added_per_sync",
+			Help:           "Number of endpoints added on each Service sync",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"service_name", "service_namespace"},
+	)
+	// EndpointSliceEndpointsRemovedPerSync tracks the number of endpoints
+	// removed on each Service sync.
+	EndpointSliceEndpointsRemovedPerSync = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Subsystem:      EndpointSliceSubsystem,
+			Name:           "endpoints_removed_per_sync",
+			Help:           "Number of endpoints removed on each Service sync",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"service_name", "service_namespace"},
+	)
+	// EndpointSliceEndpointsDesired tracks the total number of desired endpoints.
+	EndpointSliceEndpointsDesired = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Subsystem:      EndpointSliceSubsystem,
+			Name:           "endpoints_desired",
+			Help:           "Number of endpoints desired for a given Service",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"service_name", "service_namespace"},
+	)
+	// EndpointSliceSize tracks the number of endpoints in each Endpoint Slice.
+	EndpointSliceSize = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Subsystem:      EndpointSliceSubsystem,
+			Name:           "endpoint_slice_size",
+			Help:           "Number of endpoints in each Endpoint Slice",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"service_name", "service_namespace", "endpoint_slice_name"},
+	)
+	// EndpointSliceChanges tracks the number of changes to Endpoint Slices.
+	EndpointSliceChanges = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Subsystem:      EndpointSliceSubsystem,
+			Name:           "endpoint_slice_changes",
+			Help:           "The number of EndpointSlice changes",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"service_name", "service_namespace", "operation"},
```

Comment:
Generally speaking, you cannot put user input-derived fields in a metric, as the # of metrics will potentially grow w/out bound, not to mention if they are stored in internal systems, they would constitute PIIl.